File: src/csharp/Intel/Iced.UnitTests/Intel/InstructionInfoTests/RegisterInfoTestReader.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if INSTR_INFO
using System;
using System.Diagnostics;
using System.IO;
using System.Text;
using Iced.Intel;

namespace Iced.UnitTests.Intel.InstructionInfoTests {
	static class RegisterInfoTestReader {
		static readonly char[] commaSeparator = new char[] { ',' };
		static readonly char[] spaceSeparator = new char[] { ' ' };

		public static RegisterInfoTestCase[] GetTestCases() {
			var result = new RegisterInfoTestCase[IcedConstants.RegisterEnumCount];

			var filename = PathUtils.GetTestTextFilename("RegisterInfo.txt", "InstructionInfo");
			Debug.Assert(File.Exists(filename));
			int lineNo = 0;
			foreach (var line in File.ReadLines(filename)) {
				lineNo++;
				if (line.Length == 0 || line[0] == '#')
					continue;

				RegisterInfoTestCase tc;
				try {
					tc = ParseLine(line, lineNo);
				}
				catch (Exception ex) {
					throw new Exception($"Invalid line {lineNo} ({filename}): {ex.Message}");
				}
				if (result[(int)tc.Register] is not null)
					throw new InvalidOperationException($"Duplicate test, {filename}, line {lineNo}");
				result[(int)tc.Register] = tc;
			}

			var sb = new StringBuilder();
			for (int i = 0; i < result.Length; i++) {
				var tc = result[i];
				if (tc is null) {
					if (sb.Length > 0)
						sb.Append(", ");
					sb.Append(((Register)i).ToString());
				}
			}
			if (sb.Length != 0)
				throw new InvalidOperationException($"Missing tests in {filename}: " + sb.ToString());
			return result!;
		}

		static RegisterInfoTestCase ParseLine(string line, int lineNo) {
			Static.Assert(MiscInstrInfoTestConstants.RegisterElemsPerLine == 7 ? 0 : -1);
			var elems = line.Split(commaSeparator, MiscInstrInfoTestConstants.RegisterElemsPerLine);
			if (elems.Length != MiscInstrInfoTestConstants.RegisterElemsPerLine)
				throw new Exception($"Expected {MiscInstrInfoTestConstants.RegisterElemsPerLine - 1} commas");

			var tc = new RegisterInfoTestCase();
			tc.LineNumber = lineNo;
			tc.Register = ToEnumConverter.GetRegister(elems[0].Trim());
			tc.Number = NumberConverter.ToInt32(elems[1].Trim());
			tc.BaseRegister = ToEnumConverter.GetRegister(elems[2].Trim());
			tc.FullRegister = ToEnumConverter.GetRegister(elems[3].Trim());
			tc.FullRegister32 = ToEnumConverter.GetRegister(elems[4].Trim());
			tc.Size = NumberConverter.ToInt32(elems[5].Trim());
			foreach (var value in elems[6].Split(spaceSeparator, StringSplitOptions.RemoveEmptyEntries)) {
				if (!InstructionInfoDicts.RegisterFlagsTable.TryGetValue(value, out var flags))
					throw new InvalidOperationException($"Invalid flags value: {value}");
				tc.Flags |= flags;
			}
			return tc;
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/info/RegisterInfoTestReader.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.info;

import java.util.ArrayList;

import com.github.icedland.iced.x86.FileUtils;
import com.github.icedland.iced.x86.NumberConverter;
import com.github.icedland.iced.x86.PathUtils;
import com.github.icedland.iced.x86.ToRegister;
import com.github.icedland.iced.x86.internal.IcedConstants;

final class RegisterInfoTestReader {
	public static RegisterInfoTestCase[] getTestCases() {
		String filename = PathUtils.getTestTextFilename("InstructionInfo", "RegisterInfo.txt");
		ArrayList<String> lines = FileUtils.readAllLines(filename);
		RegisterInfoTestCase[] result = new RegisterInfoTestCase[IcedConstants.REGISTER_ENUM_COUNT];
		int lineNo = 0;
		for (String line : lines) {
			lineNo++;
			if (line.length() == 0 || line.charAt(0) == '#')
				continue;

			RegisterInfoTestCase tc;
			try {
				tc = parseLine(line, lineNo);
			}
			catch (Exception ex) {
				throw new UnsupportedOperationException(String.format("Invalid line %d (%s): %s", lineNo, filename, ex.getMessage()));
			}
			if (result[tc.register] != null)
				throw new UnsupportedOperationException(String.format("Duplicate test, %s, line %d", filename, lineNo));
			result[tc.register] = tc;
		}

		StringBuilder sb = new StringBuilder();
		for (int i = 0; i < result.length; i++) {
			RegisterInfoTestCase tc = result[i];
			if (tc == null) {
				if (sb.length() > 0)
					sb.append(", ");
				sb.append(i);
			}
		}
		if (sb.length() != 0)
			throw new UnsupportedOperationException(String.format("Missing tests in %s: %s", filename, sb.toString()));
		return result;
	}

	private static RegisterInfoTestCase parseLine(String line, int lineNo) {
		int expValue = 7;
		if (MiscInstrInfoTestConstants.REGISTER_ELEMS_PER_LINE != expValue)
			throw new UnsupportedOperationException();
		String[] elems = line.split(",", MiscInstrInfoTestConstants.REGISTER_ELEMS_PER_LINE);
		if (elems.length != MiscInstrInfoTestConstants.REGISTER_ELEMS_PER_LINE)
			throw new UnsupportedOperationException(String.format("Expected %d commas", MiscInstrInfoTestConstants.REGISTER_ELEMS_PER_LINE - 1));

		RegisterInfoTestCase tc = new RegisterInfoTestCase();
		tc.lineNumber = lineNo;
		tc.register = ToRegister.get(elems[0].trim());
		tc.number = NumberConverter.toInt32(elems[1].trim());
		tc.baseRegister = ToRegister.get(elems[2].trim());
		tc.fullRegister = ToRegister.get(elems[3].trim());
		tc.fullRegister32 = ToRegister.get(elems[4].trim());
		tc.size = NumberConverter.toInt32(elems[5].trim());
		for (String value : elems[6].split(" ")) {
			if (value.equals(""))
				continue;
			Integer flags = InstrInfoDicts.toRegisterFlags.get(value);
			if (flags == null)
				throw new UnsupportedOperationException(String.format("Invalid flags value: %s", value));
			tc.flags |= flags;
		}
		return tc;
	}
}
